### ISSUE TITLE
私有网关协议的frame_mod是否支持运行中更换

### ISSUE THREAD

**Issue by helloyan** [emqx/emqx#11080]
### What would you like to be added or enhanced?

查阅目前的emqx_gateway应用，里面的frame_mod在网关初始化时就定义好了，在运行中是无法更换的，希望可以增加update的函数

### Why is this needed?

不同类型的设备在认证通过后，希望加载不同的frame_mod进行报文解析

**Comment by zmstone** [comment]
Hi @helloyan 

Could you elaborate what "私有网关协议“ is?

This is unlikely something we can prioritize in the near future unless we can see this feature benefit a lot of other users.
You are welcome to send a pull request though.

**Comment by helloyan** [comment]
"私有网关协议"refers to a gateway protocol plugin that I developed based on the specifications of emqx.

I assume that there are very few people developing their own gateway plugins, so I cannot confirm if there will be more users benefiting from it. Due to the high level of coupling between gateway modules, I currently cannot fully understand their interrelationships.

**Comment by lafirest** [comment]
@helloyan 
In general, a server will load all required modules when bootstrap, hot-swap modules are usually for code patch, not for logic branch switch. For your situation, you can write a top-level frame module as the entry, and then select a corresponding sub-module by your device information, the `update` is unnecessary.

**Comment by helloyan** [comment]
It is not hot-swap, but more like using gen_server's enter_loop/5 to switch and load another module.

If I write the top-level frame module separately, do I need to rewrite the channel/frame  behaviour modules and emqx_gateway_conn? Because the loading and processing processes are handled internally by several behaviours, I hope to achieve non-intrusive development implementation.

**Comment by zmstone** [comment]
Thank you guys for the clarification and discussion.
I've added a "Feature" label to this issue based on my understanding:

What @helloyan wanted to achieve is to avoid having to create a new gateway (which is maybe what‘s @HJianBo is suggesting in the above comment), but only to replace the frame parsing module which is currently hardcoded for each different gateway.

Take stomp for example. If `frame_mod` [here](https://github.com/emqx/emqx/blob/1dfbb7ac8064b4fec88ed2858be0cce0a3aa3196/apps/emqx_gateway_stomp/src/emqx_gateway_stomp.erl#L62)
can be taken from `Config` [here](https://github.com/emqx/emqx/blob/1dfbb7ac8064b4fec88ed2858be0cce0a3aa3196/apps/emqx_gateway_stomp/src/emqx_gateway_stomp.erl#L56)
Then all @helloyan needs to do is to implement a different frame module and continue to use other functionalities.

**Comment by helloyan** [comment]
thanks for reply @HJianBo @zmstone 

Similar to what @zmstone mentioned, I actually wrote a completely new gateway plugin following @HJianBo's suggestions, use the behaviou module. I primarily referenced the CoAP protocol because my protocol architecture design did not comply with the MQTT protocol, necessitating a remapping. The current scenario I'm facing is that I need to load different frame parsing modules based on the username provided during login. As @zmstone mentioned, these modules are fixed during the initial initialization, but I would like to be able to switch the frame module during the communication process

**Comment by HJianBo** [comment]
> **I need to load different frame parsing modules based on the username provided during login**

Yes, this is a logic that the current gateway framework does not have.

But if you want to know the client's Username, do you need to first parse the client's data stream and get it from there?

Why not store the parsed Username into the State of `frame_mod`, and then call different parsing sub functions based on its value? Of course, these functions can be distributed in the module or different from the `frame_mod`. This is not a big problem.

**Comment by helloyan** [comment]
> But if you want to know the client's Username, do you need to first parse the client's data stream and get it from there?

Indeed, it is necessary to parse the login information once. Additionally, I intend to store this username and the information of the loaded module in the database, as they cannot be directly obtained from the login data.

> Why not store the parsed Username into the State of `frame_mod`, and then call different parsing sub-functions based on its value? Of course, these functions can be distributed in the module or different from the `frame_mod`. This is not a big problem.

This is a choice, but there are two reasons that led me to choose to switch module:

1. By reading the gateway source code ,according to the gateway architecture design, the frame module should only be responsible for parsing and not for logical processing. Therefore, I think the switch frame module should be handled in the channel.
2. My frame module may be more complex, so it would be more convenient to manage it as a separate file overall.

**Comment by helloyan** [comment]
I am preparing to use MySQL for login authentication. Can the frame_module related to the username be returned during authentication? From the code I see, currently only the authentication result is returned (the relevant function should be emqx_gateway_ctx:authenticate). I initially planned to use the hook "client.connected."
